internal/virtualcircuit/create.go

```diff
@@ -0,0 +1,157 @@
+		Long:  "Creates an create-virtual-circuit for specific interconnection",
+		Example: `  # Creates a new virtual-circuit named "interconnection": 
+  metal virtual-circuit create [-i connection_id] [-p port_id] [-P <project_id> ] --n <name> [-d <description>] [-v <vlanID> ] [-t <tags> ] [-vlan <vlan> ] [-speed <speed> ]
+  
+  metal virtual-circuit create [-i connection_id] [-p port_id] [-P <project_id> ] -n <name> [-m <metro>] [-r <redundancy>] [-t "shared" ] -T <service_token_type> -v <vrfs>
+  metal virtual-circuit create [-i connection_id] [-p port_id] [-P <project_id> ] -n <name> [-m <metro>] [-r <redundancy>] [-t "shared" ] -T <service_token_type> -v <vrfs>`,
+
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.SilenceUsage = true
+			var virtualCircuitCreateInput metal.VirtualCircuitCreateInput
+
+			if vrf != "" {
+				vrfVirtualCircuitCreateInput := metal.NewVrfVirtualCircuitCreateInputWithDefaults()
+
+				if projectID != "" {
+					vrfVirtualCircuitCreateInput.SetProjectId(projectID)
+				}
+
+				if description != "" {
+					vrfVirtualCircuitCreateInput.SetDescription(description)
+				}
+
+				if name != "" {
+					vrfVirtualCircuitCreateInput.SetName(name)
+				}
+
+				if nnVlan > 2 {
+					vrfVirtualCircuitCreateInput.SetNniVlan(int32(nnVlan))
+				}
+
+				if speed > 0 {
+					vrfVirtualCircuitCreateInput.SetSpeed(int32(speed))
+				}
+
+				if len(tags) > 0 {
+					vrfVirtualCircuitCreateInput.SetTags(tags)
+				}
+
+				if subnet != "" {
+					vrfVirtualCircuitCreateInput.SetSubnet(subnet)
+				}
+
+				if peerAsn > 0 {
+					vrfVirtualCircuitCreateInput.SetPeerAsn(int32(peerAsn))
+				}
+
+				if customerIP != "" {
+					vrfVirtualCircuitCreateInput.SetCustomerIp(customerIP)
+				}
+
+				if metalIP != "" {
+					vrfVirtualCircuitCreateInput.SetMetalIp(metalIP)
+				}
+
+				if md5 != "" {
+					vrfVirtualCircuitCreateInput.SetMd5(md5)
+				}
+				virtualCircuitCreateInput = metal.VirtualCircuitCreateInput{VrfVirtualCircuitCreateInput: vrfVirtualCircuitCreateInput}
+
+			} else {
+				vlanVirtualCircuitCreateInput := metal.NewVlanVirtualCircuitCreateInput(projectID)
+
+				if description != "" {
+					vlanVirtualCircuitCreateInput.SetDescription(description)
+				}
+
+				if name != "" {
+					vlanVirtualCircuitCreateInput.SetName(name)
+				}
+
+				if nnVlan > 2 {
+					vlanVirtualCircuitCreateInput.SetNniVlan(int32(nnVlan))
+				}
+
+				if vnid != "" {
+					vlanVirtualCircuitCreateInput.SetVnid(vnid)
+				}
+
+				if speed > 0 {
+					vlanVirtualCircuitCreateInput.SetSpeed(int32(speed))
+				}
+
+				if len(tags) > 0 {
+					vlanVirtualCircuitCreateInput.SetTags(tags)
+				}
+
+				virtualCircuitCreateInput = metal.VirtualCircuitCreateInput{VlanVirtualCircuitCreateInput: vlanVirtualCircuitCreateInput}
+			}
+
+			vc, _, err := c.Service.CreateInterconnectionPortVirtualCircuit(context.Background(), connectionID, portID).VirtualCircuitCreateInput(virtualCircuitCreateInput).Execute()
```

Comment:
Bear with me here, but I think we can clean this up following this pattern:

```go
// vcParamBuilder provides the interface for building the parameters to create or update VRF or VLAN VirtualCircuits
type vcParamBuilder interface {
  SetName(string)
  SetDescription(string)
  SetSpeed(int)
  SetVnid(int)
  ...
}

// compile-time assertion that VLAN parameter types implement vcParamBuilder 
var (
 _ vcParamBuilder = (*metalgo.VlanVirtualCircuitCreateInput)(nil)
 _ vcParamBuilder = (*metalgo.VrfVirtualCircuitCreateInput)(nil)
 _ vcParamBuilder = (*metalgo.VlanVirtualCircuitUpdateInput)(nil)
 _ vcParamBuilder = (*metalgo.VrfVirtualCircuitUpdateInput)(nil)
)

// setParams is a helper for defined VC parameters from Cobra arguments
func setParams(p vcParamBuilder, name, description string, speed, vni int) {
  if name > "" { p.SetName(name) }
  if description > "" { p.SetDescription(description) }
  ...
}

...
 RunE: ...
    var vcInput = metal.VirtualCircuitCreateInput
    
    if vrf != "" {
        vrfInput = metal.NewVrfVirtualCircuitCreateInputWithDefaults()
        setParams(vrfInput)
        vrfInput.SetVrf(...) // set any VRF only things
	vcInput.VrfVirtualCircuitCreateInput = vrfInput
    } else {
        vlanInput = metal.NewVlanVirtualCircuitCreateInputWithDefaults()
        setParams(vlanInput)
        vrfInput.SetVLANSomething(...) // set any VLAN only things
        vcInput.VlanVirtualCircuitCreateInput := vlanInput
    }

    vc, _, err := c.Service.CreateInterconnectionPortVirtualCircuit(context.Background(), connectionID, portID).VirtualCircuitCreateInput(virtualCircuitCreateInput).Execute()
```

Depending on how similar the Update and Create methods (`SetDescription`, `SetName`,...) and how similar the create vs update cobra arguments are, a helper like this could reduce much of the repeated code.

Reply:
I'm pretty sure the suggestions in this thread have been addressed, so I'm going to go ahead and resolve it, but feel free to comment here and unresolve it if I missed something.